src/OrleansAzureUtils/Providers/Storage/AzureBlobStorage.cs

```diff
@@ -189,33 +189,30 @@ private static string GetBlobName(string grainType, GrainReference grainId)
                 var blob = container.GetBlockBlobReference(blobName);
                 blob.Properties.ContentType = "application/json";
 
-                var containerNotFound = false;
                 try
                 {
-                    await blob.UploadTextAsync(
-                            json,
-                            Encoding.UTF8,
-                            AccessCondition.GenerateIfMatchCondition(grainState.ETag),
-                            null,
-                            null).ConfigureAwait(false);
+                    await DoConditionalUpdate(() => blob.UploadTextAsync(json, Encoding.UTF8, AccessCondition.GenerateIfMatchCondition(grainState.ETag), null, null),
+                        blobName, grainState.ETag).ConfigureAwait(false);
                 }
                 catch (StorageException exception)
                 {
+                    if (exception.IsPreconditionFailed())
+                    {
+                        throw new InconsistentStateException($"Blob storage condition not Satisfied.  BlobName: {blobName}, Container: {container.Name}, CurrentETag: {grainState.ETag}", "Unkown", grainState.ETag, exception);
+                    }
+
                     var errorCode = exception.RequestInformation.ExtendedErrorInformation?.ErrorCode;
-                    containerNotFound = errorCode == BlobErrorCodeStrings.ContainerNotFound;
-                }
-                if (containerNotFound)
-                {
+                    if (errorCode != BlobErrorCodeStrings.ContainerNotFound)
+                    {
+                        throw;
+                    }
+
                     // if the container does not exist, create it, and make another attempt
                     if (this.Log.IsVerbose3) this.Log.Verbose3((int)AzureProviderErrorCode.AzureBlobProvider_ContainerNotFound, "Creating container: GrainType={0} Grainid={1} ETag={2} to BlobName={3} in Container={4}", grainType, grainId, grainState.ETag, blobName, container.Name);
                     await container.CreateIfNotExistsAsync().ConfigureAwait(false);
 
-                    await blob.UploadTextAsync(
-                        json,
-                        Encoding.UTF8,
-                        AccessCondition.GenerateIfMatchCondition(grainState.ETag),
-                        null,
-                        null).ConfigureAwait(false);
+                    await DoConditionalUpdate(() => blob.UploadTextAsync(json, Encoding.UTF8, AccessCondition.GenerateIfMatchCondition(grainState.ETag), null, null),
```

Comment:
yeah, I briefly consider that too, but prefer to do this in case it would recurse forever due to an issue with the service, but there reality is that it's very unlikely. I added a commit that makes it recursive.